### PR TITLE
Print node, npm and yarn versions after installation

### PR DESCRIPTION
**Pull request by havenchyk**
**Description:**
Print node, npm and yarn version as it was before https://github.com/actions/setup-node/pull/147

**Related issue:**
#90

**Check list:**
- [ ] Mark if documentation changes are required.
- [x] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -105,3 +108,17 @@ function resolveVersionInput(): string {
...
+async function printEnvDetails() {
+  // Output version of node and npm that are being used
+  await exec.exec('node', ['--version']);
+
+  // Older versions of Node don't include npm, so don't let this call fail
+  await exec.exec('npm', ['--version'], {
+    ignoreReturnCode: true
+  });
+
+  await exec.exec('yarn', ['--version'], {
```
Comment (havenchyk) [inline src/main.ts:121]: probably yarn is not need as it doesn't go with node, is that correct?

Comment (nickmccurdy) [reply src/main.ts:121]: Both GitHub's runners and corepack (if enabled) will automatically install Yarn

**Comment by IvanZosimov** [comment]
Hi, @havenchyk 👋 ! Thank you for your contribution! I'd suggest using [getExecOutput()](https://github.com/actions/toolkit/blob/main/packages/exec/src/exec.ts#L44) function instead of exec() to be able to format the information in the log and also to handle such situations when we want to notify user that there is no such util in the PATH (for instance the customer changed the PATH on the self-hosted runner and erased npm). Probably it'd be interesting to play with putting version information into the [foldable group](https://github.com/actions/toolkit/blob/main/packages/core/src/core.ts#L301-L310).

**Comment by IvanZosimov** [comment]
📟 Just a gentle ping, @havenchyk

**Comment by havenchyk** [comment]
hi guys, definitely I'll take a look: resolve conflicts today-tomorrow, thanks for reminders

**Comment by havenchyk** [comment]
@dmitry-shibanov @IvanZosimov rebased, sorry for the delay

**Comment by havenchyk** [comment]
oh, just noticed a comment about refactoring, let me address it

**Comment by havenchyk** [comment]
@IvanZosimov can you please help me with the testing of changes? I see actions require some approval to run

**Comment by dmitry-shibanov** [comment]
Hello @havenchyk. Could you please run `npm run build` and `npm run format` commands. Besides, I think we need to move yarn and npm to the try-cacth block too because it can cause issues like [this](https://github.com/actions/setup-node/issues/538) for docker images when the tool does not exist or it is not supported.

**Comment by havenchyk** [comment]
@dmitry-shibanov done. I think we need to run some action to verify what will the code do. Before I fix the tests, I know they will fail

**Review by marko-zivic-93 — APPROVED** [review]

**Comment by havenchyk** [comment]
Guys, I need help with testing :) which job should I check to understand that the version is printed correctly?

File: src/main.ts [hunk 2]
```diff
@@ -108,3 +98,47 @@ function resolveVersionInput(): string {
 
   return version;
 }
+
+async function printEnvDetailsAndSetOutput() {
+  core.startGroup('Environment details');
+  // Output version of node is being used
+  try {
```
Comment (dmitry-shibanov) [inline src/main.ts:105]: Could please create a separate function getToolVersion:
```
async function getToolVersion(tool: string, options: string[]) {
  try {
    const {stdout, stderr, exitCode} = await exec.getExecOutput(tool, options, {
      ignoreReturnCode: true,
      silent: true
    });

    if (exitCode) {
      core.info(`[warning]${stderr}`);
      return '';
    }

    return stdout;
  } catch (err) {
    return '';
  }
}
```

File: __tests__/installer.test.ts [hunk 3]
```diff
@@ -249,6 +249,21 @@ describe('setup-node', () => {
 
     let expPath = path.join(toolPath, 'bin');
 
+    expect(getExecOutputSpy).toHaveBeenCalledWith(
```
Comment (havenchyk) [inline __tests__/installer.test.ts:252]: @dmitry-shibanov don't you think such check is better?

File: src/main.ts [hunk 4]
```diff
@@ -108,3 +98,35 @@ function resolveVersionInput(): string {
...
+export async function printEnvDetailsAndSetOutput() {
+  core.startGroup('Environment details');
+
+  const promises = ['node', 'npm', 'yarn'].map(async tool => {
+    const output = await getToolVersion(tool, ['--version']);
+
+    core.setOutput(`${tool}-version`, output);
+  });
+
+  await Promise.all(promises);
```
Comment (havenchyk) [inline src/main.ts:111]: I didn't check it 😄

File: __tests__/installer.test.ts [hunk 5]
```diff
@@ -41,6 +41,7 @@ describe('setup-node', () => {
   let parseNodeVersionSpy: jest.SpyInstance;
   let isCacheActionAvailable: jest.SpyInstance;
   let getExecOutputSpy: jest.SpyInstance;
+  let execExecSpy: jest.SpyInstance;
```
[the remark this answers is not in the document]
Comment (havenchyk) [reply __tests__/installer.test.ts]: removed

File: src/main.ts (hunk 2 above)
Comment (havenchyk) [reply src/main.ts]: done (a bit different)

File: src/main.ts [hunk 6]
```diff
@@ -108,3 +98,35 @@ function resolveVersionInput(): string {
...
   return version;
 }
+
+export async function printEnvDetailsAndSetOutput() {
+  core.startGroup('Environment details');
+
+  const promises = ['node', 'npm', 'yarn'].map(async tool => {
+    const output = await getToolVersion(tool, ['--version']);
+
+    core.setOutput(`${tool}-version`, output);
```
Comment (dmitry-shibanov) [inline src/main.ts:108]: node-version set on previous steps could you please remove it from there. Besides, could you please logs for versions ?

Comment (havenchyk) [reply src/main.ts:108]: I don't fully understand your comment. Please check those changes https://github.com/actions/setup-node/pull/368/files#diff-4fab5baaca5c14d2de62d8d2fceef376ddddcc8e9509d86cfa5643f51b89ce3dL43-L53

Comment (dmitry-shibanov) [reply src/main.ts:108]: Sorry, I're right. They were removed. Could you please add core.info to print versions ?

Comment (havenchyk) [reply src/main.ts:108]: Added. Did you mean this? Should I keep that `core.setOutput(`${tool}-version`, output);`?

Comment (dmitry-shibanov) [reply src/main.ts:108]: According to the issue it's only about logging. I think for now we can remove output If we get separate issue about output we'll add it. Thank you.

Comment (havenchyk) [reply src/main.ts:108]: the previous version had `node-version` in the output, should we keep it?

Comment (dmitry-shibanov) [reply src/main.ts:108]: For node-version we should keep it.

Comment (havenchyk) [reply src/main.ts]: Done

**Comment by havenchyk** [comment]
@dmitry-shibanov please take a look once again

**Review by marko-zivic-93 — APPROVED** [review]

**Review by vsafonkin — APPROVED** [review]

File: __tests__/installer.test.ts (hunk 3 above)
Comment (Jajajjshsdhdhgekskaooqoq) [reply __tests__/installer.test.ts:252]: Thank you

File: src/main.ts (hunk 4 above)
Comment (Jajajjshsdhdhgekskaooqoq) [reply src/main.ts:115]: Thank yo

**Comment by dmitry-shibanov** [comment]
Hello @havenchyk. Thank you for contribution. Sorry for the late response.

**Pull request merged**